ompi/mca/pml/base/pml_base_select.c

```diff
@@ -385,3 +378,38 @@ mca_pml_base_pml_check_selected(const char *my_pml,
     free(remote_pml);
     return OMPI_SUCCESS;
 }
+
+int
+mca_pml_base_pml_check_selected(const char *my_pml,
+                                ompi_proc_t **procs,
+                                size_t nprocs)
+{
+    int ret = 0;
+    size_t i;
+
+    if (opal_pmix_base_async_modex) {
```

Comment:
The direct modex means that we will not exchange all data at startup, so the correct condition here is to compare `opal_pmix_collect_all_data` instead of `opal_pmix_base_async_modex`. From my vision we need ro replace the all cases for this PR to use `opal_pmix_collect_all_data`.

Reply:
I'm not sure I understand what you are suggesting. Async modex means "don't perform the fence", which means we will use the direct modex method for obtaining endpts.

Collect all data tells the fence operation to include endpt data instead of being a zero-byte barrier operation. Thus, _if_ you perform a fence (i.e., `async_modex` is false) _and_ you tell the fence not to collect the data, then you will still have to perform a direct modex to get the endpt info on first communication. You just forced a synchronizing barrier into MPI_Init. Some of the transports need that barrier even if they still have to do the direct modex.

So given that understanding, are you suggesting that this test needs to be changed because it really depends on whether or not data was exchanged, and not on whether the fence was executed? I'm just seeking clarity, not questioning the suggestion.

Reply:
My point is we shouldn't rely on `opal_pmix_base_async_modex` to identify that direct modex is on. As @rhc54 mentioned `async_modex=0` and `collect_data=0` then direct modex will be performed. So my suggestion is to use `if (!opal_pmix_collect_all_data)` to identify dmodex will be used.

Reply:
@rhc54 @karasevb Do we need to use opal_pmix_collect_all_data to decide whether to use dmodex or not?

Reply:
Let me provide you with the decision tree that guides behavior:

* if async_modex is true:
    * if collect_all_data is false, then we will _not_ perform a fence operation during MPI_Init and we need to use dmodex during add_procs
    * if collect_all_data is true, then we perform a non-blocking fence operation during MPI_Init, waiting for it to complete only at the very end of MPI_Init. In this case, all data has been collected and we do _not_ need to use dmodex

* if async_modex is false, then we _always_ perform a blocking fence operation during MPI_Init
    * if collect_all_data is false, then this is a zero-byte barrier and we need to use dmodex during add_procs
    * if collect_all_data is true, then all data will be collected during the fence and we do _not_ need to use dmodex

So the answer to your question is that there are two times we have to use dmodex:
* async_modex is true and collect_all_data is false
* async_modex is false and collect_all_data is false

Thus, checking for collect_all_data being false should suffice to tell you that dmodex is going to be used.

Reply:
Thank you @rhc54. I have updated the patch